### PR TITLE
Make classpath snapshotting configurable

### PR DESCRIPTION
**Pull request by wolfs**
With this change it is possible to exclude files from affecting up-to-date ness or the build cache key from classpaths.
This can be done per `Project` basis via
```
snapshotting {
    snapshotter(RuntimeClasspath) {
        exclude 'version-info.properties'
    }
}
```

### CODE REVIEW

File: subprojects/core/src/main/java/org/gradle/api/Project.java [hunk 1]
```diff
@@ -1669,4 +1670,10 @@
      */
     @Incubating
     SoftwareComponentContainer getComponents();
+
+    @Incubating
```
Comment (wolfs) [inline subprojects/core/src/main/java/org/gradle/api/Project.java]: Do you have a suggestion for naming? We could do something like
```
resources {
    configure(RuntimeClasspathBehavior) {
        exclude 'version-info.properties`
    }
}
```
WDYT?

**Comment by wolfs** [comment]
As discussed, we will replace this PR and do smaller, self contained PRs. The first follow up PR is #1899.

File: subprojects/core/src/main/java/org/gradle/api/Project.java (hunk 1 above)
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/api/Project.java]: Should we be more specific, like:

```groovy
resources {
	runtimeClasspath {
		ignore "*.properties"
	}
}
```

File: subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/TestTaskSnapshottingConfigurationIntegrationTest.groovy [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.testing
+
+import org.gradle.integtests.fixtures.AbstractIntegrationSpec
+import org.gradle.test.fixtures.file.TestFile
+
+class TestTaskSnapshottingConfigurationIntegrationTest extends AbstractIntegrationSpec {
```
Comment (lptr) [inline subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/TestTaskSnapshottingConfigurationIntegrationTest.groovy]: > changing the resource's content, removing the resource, or adding the resource will be ignored.

Yep, we should add this.

> you cannot ignore a class.

I think we need to be able to ignore classes. For example, `GradleVersion.class` could have the version number embedded as a constant (as opposed to read from a `.properties` file).

> you can ignore a resource in a jar or a class directory.

Yep.

> ideally that you can ignore a resource from a specific library (as opposed to ignoring it from every library).

I'm not sure how you'd identify a library in this context: we have a bunch of `File` objects to process (pointing to JAR files and class directories). The only way we could identify a library would be to expect the JAR file's name to match some pattern maybe, but that sounds problematic. I think it's safe not to do this now, and come back to this feature later if users need it.

**Pull request closed without merging**